language/resource_types.md

```diff
@@ -0,0 +1,502 @@
+
+```ruby
+autorequire(:file) do
+  reqs = []
+  self[:command].scan(file_regex) { |str|
+    reqs << str
+  }
+  reqs
+end
+```
+
+Autorequires are by far the most common, though recently puppet added support for `autobefore`, `autonotify`, and `autosubscribe`.
+
+### pre_run_check
+
+The agent will call the `pre_run_check` method for each `Puppet::Type` instance in the agent's catalog. This provides an opportunity for a resource to perform consistency checks/validation against other resources in the catalog. It differs from the `validate` method, since it is called later during catalog application, and can rely on the catalog having all generated resources.
+
+### provider features
+
+Defines a feature for the type, which allows puppet to perform additional validation on the agent at catalog application time based on the selected provider. For example the `service` type defines an `enableable` feature:
+
+```ruby
+Puppet::Type.newtype(:service) do
+  feature :enableable
+end
+```
+
+And the type specifies that the `enable` property can only be managed if the selected provider supports the `enableable` feature. This validation check is performed at catalog application time, once the provider has been resolved on the agent.
+
+```ruby
+Puppet::Type.newtype(:service) do
+  newproperty(:enable, :required_features => :enableable) do
+  end
+end
+```
+
+A **provider** indicates it supports the feature using `has_feature`:
+
+```ruby
+Puppet::Type.type(:service).provide :launchd, :parent => :base do
+  has_feature :enableable
+end
+```
+
+A type can also restrict allowed parameter/property values based on provider features. For example, the `enable` property can only be set to `mask` if the provider is `maskable`:
+
+```ruby
+Puppet::Type.newtype(:service) do
+  feature :maskable
+
+  newproperty(:enable, :required_features => :enableable) do
+    ...
+    newvalue(:mask, :event => :service_disabled, :required_features => :maskable) do
+      provider.mask
+    end
+  end
+end
+```
+
+### mixins
+
+Since puppet resource types are defined in ruby, you can mixin additional functionality. The AWS module uses this to create subclasses for different types of Route53 DNS records:
+
+```ruby
+Puppet::Type.newtype(:route53_a_record) do
+  extend PuppetX::Puppetlabs::Route53Record
+  @doc = 'Type representing a Route53 DNS record.'
+  create_properties_and_params()
+end
+```
+
+where the base module defines common parameters and properties for all `Route53Records`:
+
+```ruby
+module PuppetX
+  module Puppetlabs
+    module Route53Record
+      def create_properties_and_params
+        ensurable
+        newproperty(:zone) do
+        end
+      end
+    end
+  end
+end
+```
+
+## Parameter & Property DSL Methods
+
+The following DSL methods are commonly used for individual parameters and properties within the body of `newparam` and `newproperty` respectively:
```

Comment:
Yeah, `Puppet::Type` has 2.6k lines, and some of the code implements the DSL for creating custom types, e.g. `def self.newparam`. But it also implements methods that puppet uses to manage/introspect the type, e.g. `def self.attrclass(name)`, which wouldn't normally be implemented by someone writing a custom type. For example, this might "work" but really shouldn't:

```ruby
Puppet::Type.newtype(:foo) do
  def self.attrclass(name)
    # do terrible things
  end
end
```

I've been focused on writing methods that custom type authors would use, but I may have missed some.